### DISCUSSION TITLE
simulation doesn't launch with docker

### DISCUSSION THREAD

**Discussion by ican9595** [FieldRobotEvent/Virtual_Field_Robot_Event#37] [Q&A] [answered]
Hi everyone,

we followed your instructions to setup the docker containers and everything worked fine so far. 
When we run `./start_competition_environment.sh` container a and b are both starting but the gazebo simulation doesn't start.

Do you have any suggestions why this might be?

Thank you.

Best regards

Ibrahim

**Comment by ThijsRuigrok** [comment, marked as answer]
Hi Ibrahim,

What do you mean with "the simulaiton does not start"?
a. you do not see the gazebo simulation.
b. you see the gazebo simulation but it does not move. 

if b, you can press the play button in the bottom of the gazebo screen, and the simulation will start.

**Reply by ican9595** [reply]
Hi @ThijsRuigrok ,

thanks for your fast response.
a. We don't see the gazebo simulation.

**Reply by ican9595** [reply]
We tried to solve the problem by using `docker exec -it fre_a_container_1 /bin/bash` and it seems that ROS is not running because `rosnode list` gives us the output `bash: rosnode: command not found`

Is that the expected behaviour? 

We already tried to reinstall and reclone the whole git repo but it did not solve the problem.

**Reply by ThijsRuigrok** [reply]
When starting the competition environment, the a_container will start a rosmaster. The b_container is configured to communicate with the same ros master. 

The a_container should open the gazebo GUI, so something is wrong there, probably something is wrong with the x-forwarding of the a_container. Can you inspect the logs of the a_container. To do this, start the competition environment and run `docker logs fre_a_container_1`. Can you report the output.

One other question, what kind of system are you running on? Are you running in the VM or on a direct installation?

**Reply by ThijsRuigrok** [reply]
This problem: "rosnode: command not found" occurred because you forgot to source your ros environment in the container. When running `docker exec -it fre_a_container_1 /bin/bash` you open a new shell within the container without any profile setting. The run `source ros_entrypoint.sh`, adds all the ros stuff to your current session. Afterwards when you run `rosnode list`, you should see the active nodes.

**Reply by ican9595** [reply]
We are running it on a direct installation with Ubuntu 18.04.
This is the output I am getting when running `docker logs fre_a_container_1`:

```
xacro: in-order processing became default in ROS Melodic. You can drop the option.
Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Wrn] [GuiIface.cc:199] g/gui-plugin is really loading a SystemPlugin. To load a GUI plugin please use --gui-client-plugin 
Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

Invalid MIT-MAGIC-COOKIE-1 key[Err] [RenderEngine.cc:725] Can't open display: :0
[Wrn] [RenderEngine.cc:93] Unable to create X window. Rendering will be disabled
[ INFO] [1622131223.306719570]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1622131223.308030090]: waitForService: Service [/gazebo/set_physics_properties] has not been advertised, waiting...
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 172.20.0.2
[Wrn] [RenderEngine.cc:293] Cannot initialize render engine since render path type is NONE. Ignore this warning ifrendering has been turned off on purpose.
[ INFO] [1622131223.318221796]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1622131223.319671657]: waitForService: Service [/gazebo_gui/set_physics_properties] has not been advertised, waiting...
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 172.20.0.2
[Wrn] [GuiIface.cc:297] Couldn't locate specified .ini. Creating file at "/root/.gazebo/gui.ini"
Invalid MIT-MAGIC-COOKIE-1 key[Err] [RenderEngine.cc:725] Can't open display: :0
[Wrn] [RenderEngine.cc:93] Unable to create X window. Rendering will be disabled
[Wrn] [RenderEngine.cc:293] Cannot initialize render engine since render path type is NONE. Ignore this warning ifrendering has been turned off on purpose.
[Wrn] [GuiIface.cc:120] QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
Invalid MIT-MAGIC-COOKIE-1 key[Wrn] [GuiIface.cc:120] QXcbConnection: Could not connect to display :0
[Err] [GuiIface.cc:124] Could not connect to any X display.
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
terminate called recursively
Aborted (core dumped)
[gazebo_gui-8] process has died [pid 137, exit code 134, cmd /opt/ros/melodic/lib/gazebo_ros/gzclient --verbose __name:=gazebo_gui __log:=/root/.ros/log/a3d42872-bf04-11eb-972c-0242ac140002/gazebo_gui-8.log].
log file: /root/.ros/log/a3d42872-bf04-11eb-972c-0242ac140002/gazebo_gui-8*.log
Invalid MIT-MAGIC-COOKIE-1 keyxcb_connection_has_error() returned true
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
AL lib: (EE) ALCplaybackAlsa_open: Could not open playback device 'default': No such file or directory
[Err] [OpenAL.cc:84] Unable to open audio device[default]
 Audio will be disabled.
[Err] [World.cc:2250] Unable to find model[ground_plane]
[ INFO] [1622131226.085182861]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[ INFO] [1622131226.100067146]: Physics dynamic reconfigure ready.
Warning [parser_urdf.cc:1115] multiple inconsistent <gravity> exists due to fixed joint reduction overwriting previous value [true] with [false].
Warning [parser_urdf.cc:1115] multiple inconsistent <gravity> exists due to fixed joint reduction overwriting previous value [false] with [true].
[Err] [GpuRaySensor.cc:133] Unable to create GpuRaySensor. Rendering is disabled.
[Wrn] [msgs.cc:1852] Conversion of sensor type[gpu_ray] not supported.
[ERROR] [1622131226.548677756]: No p gain specified for pid.  Namespace: /gazebo_ros_control/pid_gains/front_left_wheel
[ERROR] [1622131226.549203790]: No p gain specified for pid.  Namespace: /gazebo_ros_control/pid_gains/front_right_wheel
[ERROR] [1622131226.549697157]: No p gain specified for pid.  Namespace: /gazebo_ros_control/pid_gains/rear_left_wheel
[ERROR] [1622131226.550210991]: No p gain specified for pid.  Namespace: /gazebo_ros_control/pid_gains/rear_right_wheel
```

**Reply by ThijsRuigrok** [reply]
This is a hard one. Something is definitely wrong with the x-forwarding. Can you try to run `xhost +local:` before starting the competition environment. This might solve problems related to authorization of the docker container to the x-server.

**Reply by Rick-v-E** [reply]
I can confirm this error. Have the exact same problem on Ubuntu 20.04 on a dual boot installation. Running `xhost +local:` before starting the environment shows no improvement.

**Reply by Rick-v-E** [reply]
My display has number 1. After setting `DISPLAY:=1`, I got the following error:
```commandline
docker logs fre_a_container_1 
xacro: in-order processing became default in ROS Melodic. You can drop the option.
Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Wrn] [GuiIface.cc:199] g/gui-plugin is really loading a SystemPlugin. To load a GUI plugin please use --gui-client-plugin 
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  29
  Current serial number in output stream:  28
[Wrn] [GuiIface.cc:297] Couldn't locate specified .ini. Creating file at "/root/.gazebo/gui.ini"
[ INFO] [1622147446.138453390]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1622147446.139870563]: waitForService: Service [/gazebo_gui/set_physics_properties] has not been advertised, waiting...
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 172.23.0.2
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  29
  Current serial number in output stream:  28
Aborted (core dumped)
[gazebo-7] process has died [pid 127, exit code 134, cmd /opt/ros/melodic/lib/gazebo_ros/gzserver -u --verbose -e ode /catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/worlds/generated.world __name:=gazebo __log:=/root/.ros/log/69575cd4-bf2a-11eb-9101-0242ac170002/gazebo-7.log].
log file: /root/.ros/log/69575cd4-bf2a-11eb-9101-0242ac170002/gazebo-7*.log
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
Aborted (core dumped)
[gazebo_gui-8] process has died [pid 139, exit code 134, cmd /opt/ros/melodic/lib/gazebo_ros/gzclient --verbose __name:=gazebo_gui __log:=/root/.ros/log/69575cd4-bf2a-11eb-9101-0242ac170002/gazebo_gui-8.log].
log file: /root/.ros/log/69575cd4-bf2a-11eb-9101-0242ac170002/gazebo_gui-8*.log
```

**Reply by ThijsRuigrok** [reply]
In your host computer can you echo the display parameter `echo $DISPLAY`. If the number returned is not a zero this is a "problem". You can fix this problem by exporting the correct number to the environment in the a _container. You can do this by editing the dockercompose.yml file located in the task_<nr> folders. under the a_container tag add an environment variable and give it the number given by `echo $DISPLAY`. The dockercompose.yaml file should look like this:

'''
version: '3'
services:
  a_container:
    image: fieldrobotevent/a_container_2021
    hostname: acontainer
    volumes:
      - "./worlds:/catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/worlds"
      - "./Media:/catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/Media"
      - "./map:/catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/map"
      - "./launch:/catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/launch"
      - "/tmp/.X11-unix:/tmp/.X11-unix:rw"
      - "/tmp/.docker.xauth:/tmp/.docker.xauth:rw"
    ports:
    - "8080:8080"
    - "11345:11345"
    environment:
    - DISPLAY=:1 # replace this number

  b_container:
    depends_on:
    - "a_container"
    image: b_task_1
    build: ../b_container
    hostname: bcontainer
'''

Note the `:` after the equal sign.

**Reply by ThijsRuigrok** [reply]
Hi @Rick-v-E , You new error is related to a a difference in NVidia drivers used by your computer and expected by the docker container. You can solve this by changing the driver used for you GPU. In the software and updates menu, go to additional drivers and select the X.org nouveau (open source) driver, as shown in this image https://phoenixnap.com/kb/wp-content/uploads/2021/04/additional-gpu-drivers.png. Reboot your computer and everything works. 

The downside of this solution is that you do not utilize the full power of your dedicated GPU. I will get back when I find a better solution using the full power of your GPU.

**Reply by BenedictBauerHHN** [reply]
Switching to the X.org driver did the trick. The window now shows up, but we have like 0.8 FPS in the simulation... :'-(

**Reply by ThijsRuigrok** [reply]
@ican9595 Did this also solve your problem? If yes, please mark this question as answered.

**Reply by BenedictBauerHHN** [reply]
Hey @ThijsRuigrok 

@ican9595 is on the same team as me. Somehow the problem is not fully fixed yet. It depends on the Nvidia driver. Without this, we are probably not able to test properly...

Should we open a new question for this?

**Reply by ThijsRuigrok** [reply]
@BenedictBauerHHN what is not fixed yet? can you see the simulation?

**Reply by BenedictBauerHHN** [reply]
@ThijsRuigrok we can see it but only with 0.8 FPS without nvidia driver...

**Reply by ThijsRuigrok** [reply]
@BenedictBauerHHN. Now I understand. Yes, please create a new issue.

**Reply by MoonWalkerAZ** [reply]
Hello @ThijsRuigrok , I have the similar problem that the simulation won't start. Command: `echo $DISPLAY `gives me` :1`. I have changed the dockercompose.yml file to - DISPLAY=:1 , but gazebo simulation won't start when I run: `./start_competition_environment.sh 1`. I have even changed nvidia drives to X.org, but sadly my system is very unresponsive if I do that. I have a dual boot of Ubuntu 18.04.

Here is the log if I run `docker logs fre_a_container_1`
```

`xacro: in-order processing became default in ROS Melodic. You can drop the option.
Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Wrn] [GuiIface.cc:199] g/gui-plugin is really loading a SystemPlugin. To load a GUI plugin please use --gui-client-plugin 
Gazebo multi-robot simulator, version 9.0.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[ INFO] [1622921809.914591982]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1622921809.916352669]: waitForService: Service [/gazebo_gui/set_physics_properties] has not been advertised, waiting...
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 172.19.0.2
[Wrn] [GuiIface.cc:297] Couldn't locate specified .ini. Creating file at "/root/.gazebo/gui.ini"
libGL error: No matching fbConfigs or visuals found
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  29
  Current serial number in output stream:  28
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  29
  Current serial number in output stream:  28
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
terminate called recursively
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
Aborted (core dumped)
[gazebo-7] process has died [pid 123, exit code 134, cmd /opt/ros/melodic/lib/gazebo_ros/gzserver -u --verbose -e ode /catkin/src/Virtual_Field_Robot_Event/virtual_maize_field/worlds/generated.world __name:=gazebo __log:=/root/.ros/log/50d1a932-c635-11eb-96f7-0242ac130002/gazebo-7.log].
log file: /root/.ros/log/50d1a932-c635-11eb-96f7-0242ac130002/gazebo-7*.log
Aborted (core dumped)
[gazebo_gui-8] process has died [pid 133, exit code 134, cmd /opt/ros/melodic/lib/gazebo_ros/gzclient --verbose __name:=gazebo_gui __log:=/root/.ros/log/50d1a932-c635-11eb-96f7-0242ac130002/gazebo_gui-8.log].
log file: /root/.ros/log/50d1a932-c635-11eb-96f7-0242ac130002/gazebo_gui-8*.log
[WARN] [1622921822.950135, 0.000000]: Controller Spawner couldn't find the expected controller_manager ROS interface.`
```

**Reply by ThijsRuigrok** [reply]
Are you sure switching to the X.org driver has been done successfully? To my understanding the `X Error of failed request:  GLXBadContext` means that the docker file talks in a different "dialect" to the host GL system than expected. Using the X.org driver should solve this. If you can verify that the switch to the x.org driver was done correctly, and if you rebooted, this means that we have a new problem -_-'. 

I will look into it. A quick, but less performing fix, would be to use the VMware virtual machine, and in the VMware run the docker containers. The VMware doesn't virtualize the GPU, reducing the performance, but also preventing you from experiencing these problems.

**Reply by MoonWalkerAZ** [reply]
I have tried switching to X.org driver, but the sistem becomes so laggy and unusable. If I switch to X.org I get `:0` so this is correct. Sadly the simulation won't start.

Anyway I will try to test docker on a VMware or different computer.

**Reply by ThijsRuigrok** [reply]
You can also install gazebo and the virtual maize field directly on your computer. The performance should then be better, and hardware acceleration of you GPU should be easy then. The `start_competition_environment.sh` is primarily meant to test if all your robot code is correctly implemented in the container structure.
